pytorch_lightning/callbacks/stochastic_weight_avg.py

```diff
@@ -198,68 +221,97 @@ def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningMo
+            self.transfer_weights(self._temp_model, pl_module)
+            if self._model_contains_batch_norm:
+                self._restore_batch_norm_moments()
 
     @staticmethod
     def transfer_weights(src_pl_module: "pl.LightningModule", dst_pl_module: "pl.LightningModule"):
         for src_param, dst_param in zip(src_pl_module.parameters(), dst_pl_module.parameters()):
             dst_param.detach().copy_(src_param.to(dst_param.device))
 
-    def reset_batch_norm_and_save_state(self, pl_module: "pl.LightningModule"):
-        """Adapted from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L140-L154."""
-        self.momenta = {}
+    def _update_batch_norm_moments(
+        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", store_moments: bool = True
+    ):
+        self._batch_norm_moments = {}
+
+        train_dataloader = trainer.train_dataloader
+        if train_dataloader is None:
+            # Training data not yet connected, could be in a validation sanity check
+            return
+
+        self._update_module_batch_norm_moments(
+            train_dataloader, pl_module, self._batch_norm_moments if store_moments else None
+        )
+
+    @staticmethod
+    def _update_module_batch_norm_moments(
+        data_loader: Union[DataLoader, CombinedLoader],
+        pl_module: "pl.LightningModule",
+        moment_cache: Optional[Dict[nn.Module, Any]] = None,
+    ):
+        """Adapted from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L140-L166."""
+        prev_momenta = {}
+
+        was_training = pl_module.training
+        pl_module.train()
+
         for module in pl_module.modules():
             if not isinstance(module, nn.modules.batchnorm._BatchNorm):
                 continue
+            prev_momenta[module] = module.momentum
+            if moment_cache is not None:
+                moment_cache[module] = (module.running_mean, module.running_var)
             module.running_mean = torch.zeros_like(
                 module.running_mean, device=pl_module.device, dtype=module.running_mean.dtype
             )
             module.running_var = torch.ones_like(
                 module.running_var, device=pl_module.device, dtype=module.running_var.dtype
             )
-            self.momenta[module] = module.momentum
             module.momentum = None
             module.num_batches_tracked *= 0
 
-    def reset_momenta(self):
-        """Adapted from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L164-L165."""
-        for bn_module in self.momenta:
-            bn_module.momentum = self.momenta[bn_module]
+        # Recompute mean and variance for all batch norm layers by doing a full pass over the training data
+        for batch in data_loader:
+            if isinstance(batch, (list, tuple)):
+                batch = batch[0]
+            batch = batch.to(pl_module.device)
+            pl_module(batch)
```

Comment:
This seems to be pretty expensive to do. Especially if you do it every time you validate.

Reply:
Yeah this isn't ideal. It's only done during SWA epochs and when batch normalization is used, but I'd guess that's fairly common. It seems like the right thing to do from a correctness point of view, but possibly some users would be OK with keeping the batch normalization parameters from training as they shouldn't be massively different, so maybe this behaviour could be configurable?

Reply:
I've reverted the validation changes so this can be addressed in a follow up PR.